habitat_baselines/il/trainers/pacman_trainer.py

```diff
@@ -0,0 +1,633 @@
+                            planner_img_feats,
+                            planner_actions_in,
+                            planner_action_lengths.cpu().numpy(),
+                            planner_hidden_idx,
+                            controller_img_feats,
+                            controller_actions_in,
+                            controller_action_lengths,
+                        )
+
+                        planner_logprob = F.log_softmax(planner_scores, dim=1)
+                        controller_logprob = F.log_softmax(
+                            controller_scores, dim=1
+                        )
+
+                        planner_loss = planner_loss_fn(
+                            planner_logprob,
+                            planner_actions_out[
+                                :, : planner_action_lengths.max()
+                            ]
+                            .contiguous()
+                            .view(-1, 1),
+                            planner_masks[:, : planner_action_lengths.max()]
+                            .contiguous()
+                            .view(-1, 1),
+                        )
+
+                        controller_loss = controller_loss_fn(
+                            controller_logprob,
+                            controller_outs[
+                                :, : controller_action_lengths.max()
+                            ]
+                            .contiguous()
+                            .view(-1, 1),
+                            controller_masks[
+                                :, : controller_action_lengths.max()
+                            ]
+                            .contiguous()
+                            .view(-1, 1),
+                        )
+
+                        # zero grad
+                        optim.zero_grad()
+
+                        # update metrics
+                        metrics.update(
+                            [planner_loss.item(), controller_loss.item()]
+                        )
+
+                        (planner_loss + controller_loss).backward()
+
+                        optim.step()
+
+                        (planner_loss, controller_loss) = metrics.get_stats()
+
+                        avg_p_loss += planner_loss
+                        avg_c_loss += controller_loss
+
+                        if t % config.LOG_INTERVAL == 0:
+                            logger.info("Epoch: {}".format(epoch))
+                            logger.info(metrics.get_stat_string())
+
+                            writer.add_scalar("planner loss", planner_loss, t)
+                            writer.add_scalar(
+                                "controller loss", controller_loss, t
+                            )
+
+                            metrics.dump_log()
+
+                    # Dataloader length for IterableDataset doesn't take into
+                    # account batch size for Pytorch v < 1.6.0
+                    num_batches = math.ceil(
```

Comment:
Will this logic also work for PyTorch >= 1.6.0?

Reply:
Yes, because `len(dataset)` isn't an issue. 

PyTorch >= 1.6 [internally](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L401) uses the same logic for the dataloader's `__len__` function. Because the issue is with the dataloader's length in previous versions, we're calculating it explicitly – by operating on the dataset length.

Reply:
Ah, missed that the `len` was on the dataset.